cluster/schema/schema.go

```diff
@@ -323,27 +345,57 @@ func (s *schema) addProperty(class string, v uint64, props ...*models.Property)
 func (s *schema) addTenants(class string, v uint64, req *command.AddTenantsRequest) error {
 	req.Tenants = removeNilTenants(req.Tenants)
 
-	if ok, meta, info, err := s.multiTenancyEnabled(class); !ok {
+	ok, meta, info, err := s.multiTenancyEnabled(class)
+	if !ok {
 		return err
-	} else {
-		return meta.AddTenants(s.nodeID, req, int64(info.ReplicationFactor), v)
 	}
+
+	sc, err := meta.AddTenants(s.nodeID, req, int64(info.ReplicationFactor), v)
+	if err != nil {
+		return err
+	}
+	for status, count := range sc {
+		s.shardsCount.WithLabelValues(status).Add(float64(count))
+	}
+
+	return nil
 }
 
 func (s *schema) deleteTenants(class string, v uint64, req *command.DeleteTenantsRequest) error {
-	if ok, meta, _, err := s.multiTenancyEnabled(class); !ok {
+	ok, meta, _, err := s.multiTenancyEnabled(class)
+	if !ok {
+		return err
+	}
+	sc, err := meta.DeleteTenants(req, v)
+	if err != nil {
 		return err
-	} else {
-		return meta.DeleteTenants(req, v)
 	}
+
+	for status, count := range sc {
+		s.shardsCount.WithLabelValues(status).Sub(float64(count))
+	}
+
+	return nil
 }
 
 func (s *schema) updateTenants(class string, v uint64, req *command.UpdateTenantsRequest) error {
-	if ok, meta, _, err := s.multiTenancyEnabled(class); !ok {
+	ok, meta, _, err := s.multiTenancyEnabled(class)
+	if !ok {
 		return err
-	} else {
-		return meta.UpdateTenants(s.nodeID, req, v)
 	}
+	// TODO: What happens if it returns error because of missingShards, but still some shards are updated in `s.Sharding.Physical`?
```

Comment:
can someone also check if my understanding is correct here?

Reply:
`missingShards` will be reported to error but existed shards will be updated